File: recipes/claude/15-handle-schema-mismatches-repair/typescript/src/schemaRepair.ts
import Anthropic from "@anthropic-ai/sdk";
import { z } from "zod";
import * as fs from "node:fs";

// 1. Define Zod target schema
export const ServerDeploymentSchema = z.object({
  service_name: z.string().min(1, "service_name cannot be empty"),
  port: z.number().int().min(1).max(65535),
  status: z.enum(["healthy", "degraded", "stopped"]),
  tags: z.array(z.string()).default([]),
});

export type ServerDeployment = z.infer<typeof ServerDeploymentSchema>;

// 2. Format Zod errors into dynamic feedback for Claude
export function formatZodIssues(issues: z.ZodIssue[]): string {
  return issues
    .map((issue) => {
      const path = issue.path.join(".") || "root";
      return `- Field '${path}': [${issue.code}] ${issue.message}`;
    })
    .join("\n");
}

// 3. Multi-turn schema repair loop with circuit breaker
export async function runSchemaRepairLoop(
  prompt: string,
  maxRetries: number = 3,
  client?: Anthropic
): Promise<{ record: ServerDeployment | null; messages: Anthropic.MessageParam[] }> {
  const anthropic = client || new Anthropic();
  const systemPrompt = `You are a strict data extraction engine. Output raw, valid JSON conforming to this schema:
{
  "service_name": "string",
  "port": integer between 1 and 65535,
  "status": "healthy" | "degraded" | "stopped",
  "tags": ["string"]
}
Do not include markdown code fences, backticks, or explanatory commentary. Output JSON only.`;

  const messages: Anthropic.MessageParam[] = [
    { role: "user", content: prompt },
  ];

  for (let attempt = 1; attempt <= maxRetries; attempt++) {
    console.log(`[Attempt ${attempt}/${maxRetries}] Requesting structured extraction from Claude...`);

    const response = await anthropic.messages.create({
      model: "claude-3-7-sonnet-20250219",
      max_tokens: 1024,
      system: systemPrompt,
      messages,
    });

    const responseBlock = response.content[0];
    if (responseBlock.type !== "text") {
      throw new Error(`Unexpected block type: ${responseBlock.type}`);
    }

    const responseText = responseBlock.text.trim();
    messages.push({ role: "assistant", content: responseText });

    // Step A: Parse raw JSON syntax
    let parsedJson: unknown;
    try {
      parsedJson = JSON.parse(responseText);
    } catch (syntaxError: any) {
      const errorFeedback = `JSON syntax parsing failed: ${syntaxError.message}.\nEnsure the response is valid RFC 8259 JSON without code fences or surrounding text.`;
      console.log(`  [JSON Syntax Error] ${syntaxError.message}`);
      messages.push({ role: "user", content: errorFeedback });
      continue;
    }

    // Step B: Validate with Zod
    const validationResult = ServerDeploymentSchema.safeParse(parsedJson);
    if (validationResult.success) {
      console.log(`  [Success] Record validated on attempt ${attempt}: ${validationResult.data.service_name}`);
      return { record: validationResult.data, messages };
    }

    // Format errors and append correction feedback
    const formattedErrors = formatZodIssues(validationResult.error.issues);
    const feedback = `Schema validation failed with ${validationResult.error.issues.length} error(s):\n${formattedErrors}\n\nPlease fix these specific fields and return valid JSON adhering to the schema.`;
    console.log(`  [Schema Mismatch] Found ${validationResult.error.issues.length} issue(s)`);
    messages.push({ role: "user", content: feedback });
  }

  // Circuit breaker tripped: route to Dead Letter Queue (DLQ)
  console.log(`[Circuit Breaker] Max retries (${maxRetries}) reached without schema convergence. Routing to DLQ.`);
  const dlqEntry = {
    status: "dlq_fallback",
    timestamp: new Date().toISOString(),
    prompt,
    messages,
  };
  fs.appendFileSync("dlq_events.jsonl", JSON.stringify(dlqEntry) + "\n", "utf8");

  return { record: null, messages };
}

async function main() {
  const apiKey = process.env.ANTHROPIC_API_KEY;
  if (!apiKey) {
    console.log("ANTHROPIC_API_KEY environment variable is not set.");
    console.log("Dry-run test: verifying simulated Zod issue formatter...");
    const sampleInvalid = { name: "db-replica", port: "5432", status: "invalid_status" };
    const check = ServerDeploymentSchema.safeParse(sampleInvalid);
    if (!check.success) {
      console.log("Formatted feedback sample:\n" + formatZodIssues(check.error.issues));
    }
    return;
  }

  const testPrompt = "Deploy service redis-cache on port 6379 with status healthy and tags [cache, in-memory, key-value].";
  const { record } = await runSchemaRepairLoop(testPrompt, 3);
  if (record) {
    console.log("\nFinal Validated Record:\n" + JSON.stringify(record, null, 2));
  } else {
    console.log("\nPipeline fell back to DLQ.");
  }
}

if (process.argv[1]?.endsWith("schemaRepair.ts") || process.argv[1]?.endsWith("schemaRepair.js")) {
  main().catch(console.error);
}

File: recipes/claude/15-handle-schema-mismatches-repair/python/schema_repair.py
"""
Deterministic Schema Repair Loop with Pydantic and Anthropic Messages API.
Implements dynamic feedback turns and circuit breaker fallback to DLQ.
"""

import os
import json
from typing import List, Literal, Optional, Tuple
from pydantic import BaseModel, Field, ValidationError
from anthropic import Anthropic


class ServerDeployment(BaseModel):
    """Target schema for server deployment records."""
    service_name: str = Field(..., description="Unique alphanumeric identifier of the microservice")
    port: int = Field(..., ge=1, le=65535, description="Network port between 1 and 65535")
    status: Literal["healthy", "degraded", "stopped"] = Field(..., description="Current operational health status")
    tags: List[str] = Field(default_factory=list, description="Categorization tags")


def format_validation_errors(exc: ValidationError) -> str:
    """Format Pydantic validation errors into structured feedback for Claude."""
    error_lines = []
    for err in exc.errors():
        loc = ".".join(str(p) for p in err["loc"])
        msg = err["msg"]
        err_type = err["type"]
        error_lines.append(f"- Location: '{loc}' | Error Type: {err_type} | Details: {msg}")
    return "\n".join(error_lines)


def run_schema_repair_loop(
    prompt: str,
    max_retries: int = 3,
    client: Optional[Anthropic] = None
) -> Tuple[Optional[ServerDeployment], List[dict]]:
    """
    Executes a structured extraction loop with automated schema repair.
    
    If Claude returns invalid JSON or fails schema validation, the error is appended
    as a 'user' turn and sent back to Claude. If retries exceed max_retries, the
    circuit breaker trips and routes the payload to a Dead Letter Queue (DLQ).
    """
    if client is None:
        client = Anthropic()

    system_prompt = (
        "You are a strict data extraction engine. Output raw, valid JSON conforming to this schema:\n"
        "{\n"
        '  "service_name": "string",\n'
        '  "port": integer between 1 and 65535,\n'
        '  "status": "healthy" | "degraded" | "stopped",\n'
        '  "tags": ["string"]\n'
        "}\n"
        "Do not include markdown code fences, backticks, or explanatory commentary. Output JSON only."
    )

    messages = [
        {"role": "user", "content": prompt}
    ]

    for attempt in range(1, max_retries + 1):
        print(f"[Attempt {attempt}/{max_retries}] Requesting structured extraction from Claude...")
        
        response = client.messages.create(
            model="claude-3-7-sonnet-20250219",
            max_tokens=1024,
            system=system_prompt,
            messages=messages,
        )

        response_text = response.content[0].text.strip()
        
        # Append assistant turn to conversation history
        messages.append({"role": "assistant", "content": response_text})

        # 1. Parse JSON syntax
        try:
            parsed_json = json.loads(response_text)
        except json.JSONDecodeError as json_err:
            error_feedback = (
                f"JSON syntax parsing failed: {str(json_err)}.\n"
                f"Ensure the response is valid RFC 8259 JSON with no markdown ticks or conversational text."
            )
            print(f"  [JSON Error] {error_feedback}")
            messages.append({"role": "user", "content": error_feedback})
            continue

        # 2. Validate against Pydantic schema
        try:
            record = ServerDeployment.model_validate(parsed_json)
            print(f"  [Success] Record validated on attempt {attempt}: {record.service_name}")
            return record, messages
        except ValidationError as val_err:
            formatted_errors = format_validation_errors(val_err)
            error_feedback = (
                f"Schema validation failed with {len(val_err.errors())} error(s):\n"
                f"{formatted_errors}\n\n"
                f"Please fix these specific fields and return valid JSON adhering to the schema."
            )
            print(f"  [Schema Mismatch] Found {len(val_err.errors())} error(s)")
            messages.append({"role": "user", "content": error_feedback})

    # Circuit breaker tripped: route to Dead Letter Queue (DLQ)
    print(f"[Circuit Breaker] Max retries ({max_retries}) exceeded. Routing payload to Dead Letter Queue (DLQ).")
    dlq_record = {
        "status": "dlq_fallback",
        "prompt": prompt,
        "conversation_history": messages,
    }
    with open("dlq_events.jsonl", "a", encoding="utf-8") as dlq_file:
        dlq_file.write(json.dumps(dlq_record) + "\n")

    return None, messages


def main():
    api_key = os.environ.get("ANTHROPIC_API_KEY")
    if not api_key:
        print("ANTHROPIC_API_KEY environment variable is required to execute live inference.")
        print("Dry-run test: validating simulated repair cycle logic...")
        
        # Simulated test showing how error feedback is structured
        try:
            ServerDeployment.model_validate({"name": "api-gateway", "port": "not-an-int", "status": "unknown"})
        except ValidationError as e:
            feedback = format_validation_errors(e)
            print("Successfully verified Pydantic error formatter:")
            print(feedback)
            return

    client = Anthropic(api_key=api_key)
    test_prompt = "Provision service order-processing on port 8443, tags [commerce, checkout, pci], status running and healthy."
    record, history = run_schema_repair_loop(test_prompt, max_retries=3, client=client)
    
    if record:
        print(f"\nFinal Validated Record:\n{record.model_dump_json(indent=2)}")
    else:
        print("\nPipeline failed to converge. Checked DLQ store.")


if __name__ == "__main__":
    main()
